samples/bluetooth/direct_test_mode/src/fem/nrf21540.c

```diff
@@ -434,14 +458,38 @@ int nrf21540_init(void)
 
 	k_busy_wait(NRF21540_PD_PG_TIME_US);
 
+	atomic_set(&state, NRF21540_STATE_READY);
+
+	return 0;
+}
+
+int nrf21540_powerdown(void)
+{
+	int err;
+
+	if (!atomic_cas(&state, NRF21540_STATE_READY, NRF21540_STATE_BUSY)) {
+		return -EBUSY;
+	}
+
+	/* Power-down nRF21540 */
+	err = gpio_pin_set(nrf21540_cfg.pdn.port, nrf21540_cfg.pdn.pin, 0);
+	if (err) {
+		return err;
+	}
+
+	atomic_set(&state, NRF21540_STATE_OFF);
+
 	return 0;
 }
 
 int nrf21540_antenna_select(enum nrf21540_ant ant)
 {
 	int err = 0;
+	uint32_t prev_state;
+
+	prev_state = atomic_set(&state, NRF21540_STATE_BUSY);
```

Comment:
What if the state is already Busy? Won't it disrupt other operation that set the state to busy?

I think it is safer to use CAS operation here assuming that the antenna select can only happen during the STATE_ON (powered-up).

Reply:
It won't because we also check there a previous state. If previous state was set to a busy then this function returns an error `-EBUSY`
```
	prev_state = atomic_set(&state, NRF21540_STATE_BUSY);

	if (prev_state == NRF21540_STATE_BUSY) {
		return -EBUSY;
	}
```